core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,466 @@
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.Diff;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.joda.Joda;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.Index;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
+
+/**
+ * A collection of tombstones for explicitly marking indices as deleted in the cluster state.
+ *
+ * The cluster state contains a list of index tombstones for indices that have been
+ * deleted in the cluster.  Because cluster states are processed asynchronously by
+ * nodes and a node could be removed from the cluster for a period of time, the
+ * tombstones remain in the cluster state for a fixed period of time, after which
+ * they are purged.
+ */
+final public class IndexGraveyard implements MetaData.Custom {
+
+    /**
+     * Setting for the maximum tombstones allowed in the cluster state;
+     * prevents the cluster state size from exploding too large, but it opens the
+     * very unlikely risk that if there are greater than MAX_TOMBSTONES index
+     * deletions while a node was offline, when it comes back online, it will have
+     * missed index deletions that it may need to process.
+     */
+    public static final Setting<Integer> SETTING_MAX_TOMBSTONES = Setting.intSetting("cluster.indices.tombstones.size",
+                                                                                     500, // the default maximum number of tombstones
+                                                                                     Setting.Property.NodeScope);
+
+    public static final IndexGraveyard PROTO = new IndexGraveyard(new ArrayList<>());
+    public static final String TYPE = "index-graveyard";
+    private static final ParseField TOMBSTONES_FIELD = new ParseField("tombstones");
+    private static final ObjectParser<List<Tombstone>, Void> GRAVEYARD_PARSER = new ObjectParser<>("index_graveyard", ArrayList::new);
+    static {
+        GRAVEYARD_PARSER.declareObjectArray(List::addAll, Tombstone.getParser(), TOMBSTONES_FIELD);
+    }
+
+    private final List<Tombstone> tombstones;
+
+    private IndexGraveyard(final List<Tombstone> list) {
+        tombstones = Collections.unmodifiableList(list);
```

Comment:
Its a private constructor, only called from the Builder, would an assert be more appropriate?